src/mds/MDCache.cc

```diff
@@ -12933,3 +12936,23 @@ bool MDCache::dump_inode(Formatter *f, uint64_t number) {
   f->close_section();
   return true;
 }
+
+void MDCache::handle_mdsmap(const MDSMap &mdsmap) {
+  // process export_pin_delayed_queue whenever a new MDSMap received
+  auto &q = export_pin_delayed_queue;
+  auto it = q.begin();
+  while (it != q.end()) {
+    auto cur = it++;
+    CInode *in = *cur;
+    mds_rank_t export_pin = in->get_export_pin(false);
+    dout(10) << " delayed export_pin=" << export_pin << " on " << *in 
```

Comment:
Got this SIGSEGV in qa testing:

````
2019-07-20T20:15:27.771 INFO:tasks.ceph.mds.a.smithi136.stderr:*** Caught signal (Segmentation fault) **
2019-07-20T20:15:27.772 INFO:tasks.ceph.mds.a.smithi136.stderr: in thread 7fe1b8708700 thread_name:ms_dispatch
2019-07-20T20:15:27.773 INFO:tasks.ceph.mds.a.smithi136.stderr: ceph version 15.0.0-2917-g7a175be (7a175be793b17bd7a1914aca0414200b0c349885) octopus (dev)
2019-07-20T20:15:27.774 INFO:tasks.ceph.mds.a.smithi136.stderr: 1: (()+0x12890) [0x7fe1bf0f9890]
2019-07-20T20:15:27.774 INFO:tasks.ceph.mds.a.smithi136.stderr: 2: (operator<<(std::ostream&, CInode const&)+0x175) [0x5589cc3cd9b5]
2019-07-20T20:15:27.774 INFO:tasks.ceph.mds.a.smithi136.stderr: 3: (MDCache::handle_mdsmap(MDSMap const&)+0x185) [0x5589cc2cb0c5]
2019-07-20T20:15:27.774 INFO:tasks.ceph.mds.a.smithi136.stderr: 4: (MDSRankDispatcher::handle_mds_map(boost::intrusive_ptr<MMDSMap const> const&, MDSMap const&)+0xadd) [0x5589cc1bae8d]
2019-07-20T20:15:27.774 INFO:tasks.ceph.mds.a.smithi136.stderr: 5: (MDSDaemon::handle_mds_map(boost::intrusive_ptr<MMDSMap const> const&)+0xdc4) [0x5589cc1929e4]
2019-07-20T20:15:27.774 INFO:tasks.ceph.mds.a.smithi136.stderr: 6: (MDSDaemon::handle_core_message(boost::intrusive_ptr<Message const> const&)+0xb6) [0x5589cc193fd6]
2019-07-20T20:15:27.774 INFO:tasks.ceph.mds.a.smithi136.stderr: 7: (MDSDaemon::ms_dispatch2(boost::intrusive_ptr<Message> const&)+0xbb) [0x5589cc194cab]
2019-07-20T20:15:27.775 INFO:tasks.ceph.mds.a.smithi136.stderr: 8: (DispatchQueue::entry()+0x1201) [0x7fe1bf9be2c1]
2019-07-20T20:15:27.775 INFO:tasks.ceph.mds.a.smithi136.stderr: 9: (DispatchQueue::DispatchThread::entry()+0xd) [0x7fe1bfa5774d]
2019-07-20T20:15:27.775 INFO:tasks.ceph.mds.a.smithi136.stderr: 10: (()+0x76db) [0x7fe1bf0ee6db]
2019-07-20T20:15:27.775 INFO:tasks.ceph.mds.a.smithi136.stderr: 11: (clone()+0x3f) [0x7fe1be2d488f]
````

gdb details:

````
(gdb) bt
#0  raise (sig=sig@entry=11) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00005589cc4d5190 in reraise_fatal (signum=11) at ./src/global/signal_handler.cc:81
#2  handle_fatal_signal (signum=11) at ./src/global/signal_handler.cc:326
#3  <signal handler called>
#4  0x00005589cc3cd9b5 in operator<< (out=..., in=...) at ./src/mds/CInode.cc:114
#5  0x00005589cc2cb0c5 in MDCache::handle_mdsmap (this=0x5589cedbd200, mdsmap=...) at ./src/mds/MDCache.cc:12958
#6  0x00005589cc1bae8d in MDSRankDispatcher::handle_mds_map (this=0x5589cedaf000, m=..., oldmap=...) at ./src/mds/MDSRank.cc:2398
#7  0x00005589cc1929e4 in MDSDaemon::handle_mds_map (this=this@entry=0x5589ce100900, m=...) at ./src/mds/MDSDaemon.cc:878
#8  0x00005589cc193fd6 in MDSDaemon::handle_core_message (this=this@entry=0x5589ce100900, m=...) at ./src/mds/MDSDaemon.cc:1066
#9  0x00005589cc194cab in MDSDaemon::ms_dispatch2 (this=0x5589ce100900, m=...) at ./src/mds/MDSDaemon.cc:1040
#10 0x00007fe1bf9be2c1 in DispatchQueue::entry() () from /usr/lib/ceph/libceph-common.so.0
#11 0x00007fe1bfa5774d in DispatchQueue::DispatchThread::entry() () from /usr/lib/ceph/libceph-common.so.0
#12 0x00007fe1bf0ee6db in start_thread (arg=0x7fe1b8708700) at pthread_create.c:463
#13 0x00007fe1be2d488f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
(gdb) frame 5
#5  0x00005589cc2cb0c5 in MDCache::handle_mdsmap (this=0x5589cedbd200, mdsmap=...) at ./src/mds/MDCache.cc:12958
12958   ./src/mds/MDCache.cc: No such file or directory.
(gdb) print cur
$12 = 0x5589cee61800
(gdb) print in
$13 = (CInode *) 0x5589cee61800
(gdb) print q
$14 = std::set with 1 element = {[0] = 0x5589cee61800}
````

I don't know if my code suggestions will help. Can you do more testing?

Reply:
Test that failed (one of many): http://pulpito.ceph.com/pdonnell-2019-07-20_02:06:29-multimds-wip-pdonnell-testing-20190719.231534-distro-basic-smithi/4131950/

Reply:
Yes, I will do some tests.